### PR TITLE
Remove reliance on `sshpass` and utilize `SSH_ASKPASS`

### PR DESCRIPTION
**Pull request by sivel**
##### SUMMARY

Instead of attempting to match prompts, and respond to prompts with `sshpass`, which brings along some additional complexities, rely on `SSH_ASKPASS` to respond with passwords.

##### ISSUE TYPE

- Feature Pull Request

##### ADDITIONAL INFORMATION

**Why does this use shared memory and not a pipe for communication?**

OpenSSH almost immediately on invocation closes all FDs above stderr (2):

https://github.com/openssh/openssh-portable/blob/1bc426f51b0a5cfdcfbd205218f0b6839ffe91e9/ssh.c#L688-L692

This means, that regardless of having an inhertiable pipe, without `CLOEXEC` the FDs of the pipe are closed, and are thus unusable.

Additionally, through testing various mechanisms, such as named pipes, I determined the use of shared memory was more simplisitic, and pushed concerns into the Python SharedMemory implementation, instead of needing to roll many protections myself.

##### TO DO

- [x] The location of the `ssh_askpass.py` script currently causes a sanity failure because it has a shebang and is executable. There are some ideas and concerns to explore. Because we simply use `#!/usr/bin/env python3` in the script, it could attempt to execute on a python interpreter without `SharedMemory`.  However, `SSH_ASKPASS` must be a singular command, not a command with args, so we cannot execute `f'{sys.executable} -m ansible.ssh_askpass'`. This leaves us with another alternative of exposing as an entrypoint, although we could make it "private".
- [x] `sshpass` use in ansible has been around forever (Nov 2012) and obviously works in a set of broad scenarios. Will `SSH_ASKPASS` be as reliable?  Should we keep `sshpass` support, and allow it to be toggled back on?

### CODE REVIEW

**Review by bcoca — APPROVED** [review]

File: lib/ansible/plugins/connection/ssh.py [hunk 1]
```diff
@@ -558,6 +528,24 @@ def wrapped(self: Connection, *args: P.args, **kwargs: P.kwargs) -> tuple[int, b
     return wrapped
 
 
+def _clean_shm(func):
+    def inner(self, *args, **kwargs):
+        try:
+            ret = func(self, *args, **kwargs)
+        finally:
+            with contextlib.suppress(AttributeError):
```
Comment (nitzmahone) [inline lib/ansible/plugins/connection/ssh.py]: I'm normally all about "as small as possible" for try blocks, but since it appears you don't want the rest of the cleanup code to execute if the close fails, any reason not to merge the `suppress` blocks?

File: lib/ansible/cli/adhoc.py [hunk 2]
```diff
@@ -20,6 +23,9 @@
 from ansible.playbook.play import Play
 from ansible.utils.display import Display
 
+if os.getenv('_ANSIBLE_SSH_ASKPASS_SHM'):
```
Comment (nitzmahone) [inline lib/ansible/cli/adhoc.py]: I generally love the idea of centralizing on the adhoc entrypoint (for everything :laughing: ), but if this var was ever propagated (eg, a loosely-configured SSH `SendEnv`), remote adhoc invocations wouldn't work over an SSH/Ansible connection so-configured. If we really don't want to create a new entrypoint, maybe at least ensure that no additional args were passed?

File: lib/ansible/cli/_ssh_askpass.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+def main() -> None:
+    kwargs: dict[str, t.Any] = {}
+    if sys.version_info[:2] >= (3, 13):
+        kwargs['track'] = False
+    try:
+        shm = SharedMemory(name=os.environ['_ANSIBLE_SSH_ASKPASS_SHM'], **kwargs)
+    except FileNotFoundError:
+        # We must be running after the ansible fork is shutting down
+        sys.exit(1)
+    sys.stdout.buffer.write(shm.buf.tobytes().rstrip(b'\x00'))
```
Comment (nitzmahone) [inline lib/ansible/cli/_ssh_askpass.py]: Explicit `flush`, just in case there's a weird problem shutting down the SharedMemory instance? It's probably unnecessary, but also probably couldn't hurt.

File: lib/ansible/plugins/connection/ssh.py [hunk 4]
```diff
@@ -906,6 +858,43 @@ def _examine_output(self, source: str, state: str, b_chunk: bytes, sudoable: boo
...
+
+        conn_password = self.get_option('password') or self._play_context.password
+        pkcs11_provider = self.get_option("pkcs11_provider")
+        if not conn_password and pkcs11_provider:
+            raise AnsibleError("to use pkcs11_provider you must specify a password/pin")
+
+        if not conn_password:
+            return env, popen_kwargs
+
+        b_conn_password = conn_password.encode('utf-8')
```
Comment (nitzmahone) [inline lib/ansible/plugins/connection/ssh.py]: It seems like there should be a definitive "I won't be using a password" signal somewhere that could explicitly prevent lighting this up.

File: lib/ansible/plugins/connection/ssh.py [hunk 5]
```diff
@@ -906,6 +858,43 @@ def _examine_output(self, source: str, state: str, b_chunk: bytes, sudoable: boo
...
+            # deprecated: description='track argument for SharedMemory always available' python_version='3.13'
+            kwargs['track'] = False
+        self.shm = shm = SharedMemory(create=True, size=16384, **kwargs)  # type: ignore[arg-type]
+
+        shm.buf[:len(b_conn_password)] = bytearray(b_conn_password)
+        shm.close()
+
+        env['_ANSIBLE_SSH_ASKPASS_SHM'] = str(self.shm.name)
+        adhoc = pathlib.Path(sys.argv[0]).with_name('ansible')
+        env['SSH_ASKPASS'] = str(adhoc) if adhoc.is_file() else 'ansible'
```
Comment (nitzmahone) [inline lib/ansible/plugins/connection/ssh.py:976]: Like above, if for some reason someone was using a non-password or custom config that needed one of these envvars for something else, it'd probably be good to be able to explicitly avoid setting them. `DISPLAY` in particular seems like one that someone might actually want to be able to pass through for wacky X11 remoting stuff.

**Review by nitzmahone — COMMENTED** [review]
I like it- maybe we should do it alongside the sshpass support and switch it to the default?

**Review by nitzmahone — COMMENTED** [review]
Yeah, I'm +1 for making this the default whenever it merges, but I think we should probably keep the sshpass support around for a bit (maybe run it through a normal deprecation cycle?)

**Comment by bcoca** [comment]
I would still default to sshpass and use this when it is not found, adding switch to default and deprecation of sshpass usage in next release.

File: lib/ansible/plugins/connection/ssh.py [hunk 6]
```diff
@@ -610,24 +599,6 @@ def _create_control_path(
         cpath = '%(directory)s/' + digest[:10]
         return cpath
 
-    @staticmethod
-    def _sshpass_available() -> bool:
-        global SSHPASS_AVAILABLE
```
Comment (bcoca) [inline lib/ansible/plugins/connection/ssh.py:616]: if we restore the sshpass code, a better 'available' test would be to use the get_bin_path function

File: lib/ansible/plugins/connection/ssh.py (hunk 1 above)
Comment (sivel) [reply lib/ansible/plugins/connection/ssh.py]: I had to re-read a few times to remember. But the first could be switched to `if self.shm:` instead of catching an exception. I've done that, because it's actually faster than handling an exception for the common case.

The 2nd needs to say, but they cannot be combined.  We still need to run the code within `if not _HAS_RESOURCE_TRACK:` regardless of the `FileNotFoundError` but only when `self.shm` is set.

File: lib/ansible/cli/adhoc.py [hunk 7]
```diff
@@ -20,6 +24,13 @@
 from ansible.playbook.play import Play
 from ansible.utils.display import Display
 
+# as the SSH_ASKPASS script can only be invoked as a singular command
+# python -m doesn't work here, so assume that if we are executing
+# and there is only a single item in argv plus the executable, and the
+# env var is set we are in SSH_ASKPASS mode
+if sys.argv[2:] == [] and os.getenv('_ANSIBLE_SSH_ASKPASS_SHM'):
+    _ssh_askpass.main()
+
```
Comment (sivel) [inline lib/ansible/cli/adhoc.py]: I don't really love this check.  The script is invoked with a single arg that can look like numerous things, including:

* `user@host's password: `
* `Enter PIN for 'token': `

Both of the above are things that we know we support today.  There may be others.

This means that we assume that any argv of exactly 2 elements like `['ansible', 'anything']` plus the `_ANSIBLE_SSH_ASKPASS_SHM` env var means it's safe to assume we are meaning to run as an askpass program.

This really only impacts invalid uses of the adhoc command, or `ansible --help`, so although I don't like it, the potential for incorrect behavior is low.

File: lib/ansible/plugins/connection/ssh.py [hunk 8]
```diff
@@ -62,6 +62,21 @@
...
               - name: ansible_ssh_pass
               - name: ansible_ssh_password
+      password_mechanism:
+          description: Mechanism to use for handling ssh password prompt
+          type: string
+          default: ssh_askpass
+          choices:
+              - ssh_askpass
+              - sshpass
+              - disable
```
Comment (sivel) [inline lib/ansible/plugins/connection/ssh.py:72]: Should there be an `auto`? That if `sshpass` is available, will be chosen, and fallback to `ssh_askpass`?

File: lib/ansible/cli/adhoc.py [hunk 9]
```diff
@@ -20,6 +24,13 @@
 from ansible.playbook.play import Play
 from ansible.utils.display import Display
 
+# as the SSH_ASKPASS script can only be invoked as a singular command
+# python -m doesn't work here, so assume that if we are executing
+# and there is only a single item in argv plus the executable, and the
+# env var is set we are in SSH_ASKPASS mode
+if sys.argv[2:] == [] and os.getenv('_ANSIBLE_SSH_ASKPASS_SHM'):
+    _ssh_askpass.main()
```
Comment (mattclay) [inline lib/ansible/cli/adhoc.py]: Since the `_ssh_askpass` module doesn't use any `ansible` imports, it may be a good idea to avoid the usual ansible startup overhead by inserting this block of code very early in `lib/ansible/cli/__init__.py`, before any of the `ansible` imports.

The conditional could be updated to accommodate this:

`if 1 <= len(sys.argv) <= 2 and os.path.basename(sys.argv[0]) == "ansible" and os.getenv('_ANSIBLE_SSH_ASKPASS_SHM'):`

Comment (sivel) [reply lib/ansible/cli/adhoc.py]: I know it's not really a primary concern, but hard coding the `argv[0] == "ansible"` means it wouldn't work in fallible.

Comment (mattclay) [reply lib/ansible/cli/adhoc.py]: Isn't that already the case because of the following lines?

https://github.com/ansible/ansible/blob/2c7feb4b5c1082c4b382e2885be3333fc87a3345/lib/ansible/plugins/connection/ssh.py#L971-L972

Comment (sivel) [reply lib/ansible/cli/adhoc.py]: Haha, don't point out inconsistencies in my arguments!

File: test/integration/targets/connection_ssh/test_ssh_askpass.yml [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+    - service:
+        name: ssh{{ '' if ansible_facts.os_family == 'Debian' else 'd' }}
+        state: restarted
+      when: ansible_facts.system != 'Darwin'
+
+    - command:
+        argv:
+          - ansible
+          - localhost
```
Comment (mattclay) [inline test/integration/targets/connection_ssh/test_ssh_askpass.yml:29]: This should connect to `testhost`, not `localhost`, since they're not always the same.

```suggestion
          - testhost
```

Comment (sivel) [reply test/integration/targets/connection_ssh/test_ssh_askpass.yml:29]: I actually had that, by using `../../inventory`, the problem is that the inventory doesn't specify `ansible_host=127.0.0.1`, and as such `testhost` isn't resolvable:

```
ssh: Could not resolve hostname testhost: Name or service not known
```

There is additionally a `test_connection.inventory` that provides 2 hosts, which are for pipelining tests that do set `ansible_host=localhost`.

File: lib/ansible/cli/adhoc.py (hunk 9 above)
Comment (sivel) [reply lib/ansible/cli/adhoc.py]: Moved

File: lib/ansible/plugins/connection/ssh.py (hunk 8 above)
Comment (sivel) [reply lib/ansible/plugins/connection/ssh.py:72]: As it stands now, we have more in favor of switching to `ssh_askpass` as the default.  Assuming we keep it this way, I'll need to write a porting guide entry.

File: test/integration/targets/connection_ssh/test_ssh_askpass.yml (hunk 10 above)
Comment (mattclay) [reply test/integration/targets/connection_ssh/test_ssh_askpass.yml:29]: Ah yes, I'd forgotten that connection tests run in the controller context, so `testhost` is `localhost` in this case.

File: test/integration/targets/connection_ssh/test_ssh_askpass.yml [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+          - -m
+          - command
+          - -a
+          - id
+          - -vvv
+          - -e
+          - ansible_connection=ssh
```
Comment (mattclay) [inline test/integration/targets/connection_ssh/test_ssh_askpass.yml:38]: I'd add `ansible_pipelining=yes` here to speed up the test and reduce output.

Comment (sivel) [reply test/integration/targets/connection_ssh/test_ssh_askpass.yml:38]: added

File: lib/ansible/cli/_ssh_askpass.py [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+from __future__ import annotations
+
+import os
+import sys
+import typing as t
+from multiprocessing.shared_memory import SharedMemory
+
+
+def main() -> t.NoReturn:
```
Comment (webknjaz) [inline lib/ansible/cli/_ssh_askpass.py]: (newer alias)
```suggestion
def main() -> t.Never:
```

File: lib/ansible/cli/_ssh_askpass.py [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+    kwargs: dict[str, t.Any] = {}
```
Comment (webknjaz) [inline lib/ansible/cli/_ssh_askpass.py]: Looks like this doesn't really need any:
```suggestion
    kwargs: dict[str, bool] = {}
```
Though, MyPy will probably demand
```suggestion
    kwargs: dict[str, bool] | dict[t.Never, t.Never] = {}
```
*(it's probably better to use a typed dict, still)*

File: lib/ansible/plugins/connection/ssh.py [hunk 14]
```diff
@@ -928,39 +1009,30 @@ def _bare_run(self, cmd: list[bytes], in_data: bytes | None, sudoable: bool = Tr
         else:
             cmd = list(map(to_bytes, cmd))
 
-        conn_password = self.get_option('password') or self._play_context.password
+        env, popen_kwargs = self._init_shm()
```
Comment (jborean93) [inline lib/ansible/plugins/connection/ssh.py]: Looks like you aren't using `env` directly anymore and it is part of the `popen_kwargs`.  This should simplify the return value for `self._init_shm()`.

File: lib/ansible/plugins/connection/ssh.py [hunk 15]
```diff
@@ -1179,10 +1251,11 @@ def _bare_run(self, cmd: list[bytes], in_data: bytes | None, sudoable: bool = Tr
             p.stdout.close()
             p.stderr.close()
 
-        if self.get_option('host_key_checking'):
-            if cmd[0] == b"sshpass" and p.returncode == 6:
-                raise AnsibleError('Using a SSH password instead of a key is not possible because Host Key checking is enabled and sshpass does not support '
-                                   'this.  Please add this host\'s fingerprint to your known_hosts file to manage this host.')
+        conn_password = self.get_option('password') or self._play_context.password
```
Comment (jborean93) [inline lib/ansible/plugins/connection/ssh.py:1257]: Is the play_context.password needed anymore? I thought the `password` option name would have it set up automatically for us.

File: lib/ansible/plugins/connection/ssh.py [hunk 16]
```diff
@@ -907,6 +943,48 @@ def _examine_output(self, source: str, state: str, b_chunk: bytes, sudoable: boo
...
         return b''.join(output), remainder
 
+    def _init_shm(self) -> tuple[dict[str, str], dict[str, t.Any]]:
+        env = os.environ.copy()
+        popen_kwargs: dict[str, t.Any] = {}
+
+        if self.get_option('password_mechanism') != 'ssh_askpass':
+            return env, popen_kwargs
+
+        conn_password = self.get_option('password') or self._play_context.password
```
Comment (jborean93) [inline lib/ansible/plugins/connection/ssh.py:954]: Same here, is there a reason we are still checking the play_context?

File: lib/ansible/plugins/connection/ssh.py [hunk 17]
```diff
@@ -1179,10 +1251,11 @@ def _bare_run(self, cmd: list[bytes], in_data: bytes | None, sudoable: bool = Tr
...
             p.stderr.close()
 
-        if self.get_option('host_key_checking'):
-            if cmd[0] == b"sshpass" and p.returncode == 6:
-                raise AnsibleError('Using a SSH password instead of a key is not possible because Host Key checking is enabled and sshpass does not support '
-                                   'this.  Please add this host\'s fingerprint to your known_hosts file to manage this host.')
+        conn_password = self.get_option('password') or self._play_context.password
+        hostkey_fail = (cmd[0] == b"sshpass" and p.returncode == 6) or b"read_passphrase: can't open /dev/tty" in b_stderr
+        if password_mechanism and self.get_option('host_key_checking') and conn_password and hostkey_fail:
+            raise AnsibleError('Using a SSH password instead of a key is not possible because Host Key checking is enabled. '
```
Comment (jborean93) [inline lib/ansible/plugins/connection/ssh.py:1264]: It looks like SSH will error before hitting this branch. In this scenario the fingerprint prompt is sent to the `SSH_ASKPASS` program which will output something that isn't `yes/no/fingerprint` which causes it to re-run the program and then eventually fail.

When printing the argv to our sshpass shim for this scenario I get the following args (surrounded by `[]`). It also calls it twice, the first for the fingerprint prompt then the second to clarify before it ends up failing.

```
[/home/jborean/.pyenv/versions/ansible-313/bin/ansible], [The authenticity of host 'dc01.domain.test (fd7c:dcb6:1e61:7496:2df9:f75a:65:e71d)' can't be established.
ED25519 key fingerprint is SHA256:zVjz7SfuxlpadnRhzIlFXO4pmuw+TsKz9gqlidyzhjQ.
This host key is known by the following other names/addresses:
    ~/.ssh/known_hosts:4: server2025.domain.test
    ~/.ssh/known_hosts:7: ddc01.domain.test
Are you sure you want to continue connecting (yes/no/[fingerprint])? ]

[/home/jborean/.pyenv/versions/ansible-313/bin/ansible], [Please type 'yes', 'no' or the fingerprint: ]
```

**Review by jborean93 — CHANGES_REQUESTED** [review]

File: lib/ansible/plugins/connection/ssh.py (hunk 14 above)
Comment (sivel) [reply lib/ansible/plugins/connection/ssh.py]: Addressed

File: lib/ansible/plugins/connection/ssh.py (hunk 15 above)
Comment (sivel) [reply lib/ansible/plugins/connection/ssh.py:1257]: I honestly don't know. It was copied from https://github.com/ansible/ansible/pull/83936/files#diff-38cec806ea1a1ee7c3a286c7865334ecfba7ccc49e21d4e8fb8ec1b17938fda6L680 which I did not modify.

One of the other lines was just moved.

File: lib/ansible/plugins/connection/ssh.py (hunk 17 above)
Comment (sivel) [reply lib/ansible/plugins/connection/ssh.py:1264]: That's...fun? I'll have to think this over a little bit. Since we aren't just blindly passing the message through to a user to read, understand, and respond to, it makes it somewhat complicated to handle this.

I'm not currently convinced that we can rely on matching the message in argv. I'd have to evaluate whether we can force the locale to `C`/`C.UTF-8` and whether that would impact the message passed to the script.  Looking for `fingerprint` in the arg might be sufficient.

Comment (jborean93) [reply lib/ansible/plugins/connection/ssh.py:1264]: Yea parsing the message doesn't sound ideal, I assume it's also problematic because we are clearing the password in the shared memory. Is there a way to just have SSH fail on a prompt, I recall there being an option but unsure if it was Linux/version specific. That might be a decent workaround until/if we decide to support responding to the host key prompt in our askpass implementation.

File: lib/ansible/plugins/connection/ssh.py (hunk 15 above)
Comment (jborean93) [reply lib/ansible/plugins/connection/ssh.py:1257]: Looks like it is set on the `password` option through `task_executor.py` https://github.com/ansible/ansible/blob/09391f38f009ec58b5759dbd74df34fd281ef3ac/lib/ansible/executor/task_executor.py#L1082-L1087

We can always remove the fallback in a separate PR if you just want to keep it as is for now.

Comment (sivel) [reply lib/ansible/plugins/connection/ssh.py:1257]: I think I'll leave it as is for now, for later follow up.

**Review by jborean93 — APPROVED** [review]

**Comment by webknjaz** [comment]
macOS is failing even after rerun. I noticed the new VM. It already revealed even more race conditions than before due to slowness, which @nitzmahone fixed but I imagine this might surface another one..

**Comment by sivel** [comment]
> > macOS is failing even after rerun
> 
> That failure may be somewhat legitimate. The same setup targets I am using were recently modified. I'll look into it more.

Looks like it is a change in macos 15, maybe even 15.3, preventing stopping sshd.  I'll need to evaluate a change for macos.  I tested against the previous macos/14.3 remote and the tests work as expected.

File: test/integration/targets/connection_ssh/test_ssh_askpass.yml [hunk 18]
```diff
@@ -3,22 +3,20 @@
...
-      register: sshd_confs
-
-    - lineinfile:
-        path: '{{ item }}'
-        regexp: '^PasswordAuthentication'
-        line: PasswordAuthentication yes
-      loop: '{{ sshd_confs.files|default([{"path": "/etc/ssh/sshd_config"}], true)|map(attribute="path") }}'
-
-    - when: ansible_facts.system == 'Darwin'
+    # macos currently allows password auth, and macos/15.3 prevents restarting sshd
```
Comment (webknjaz) [inline test/integration/targets/connection_ssh/test_ssh_askpass.yml:6]: Does it? The VM is supposed to grant full disk access to the administrator. Could you add more details here?

File: lib/ansible/cli/__init__.py [hunk 19]
```diff
@@ -9,6 +9,16 @@
 import os
 import sys
 
+
+# as the SSH_ASKPASS script can only be invoked as a singular command
+# python -m doesn't work here, so assume that if we are executing
```
Comment (webknjaz) [inline lib/ansible/cli/__init__.py]: > python -m doesn't work here

Could you explain this statement? Does it mean that “we don't invoke it” or “it's not functional”?

Comment (sivel) [reply lib/ansible/cli/__init__.py]:
```
# as the SSH_ASKPASS script can only be invoked as a singular command
# python -m doesn't work here
```

That means that the value specified is treated as `argv[0]`, so `python -m` isn't callable.

File: test/integration/targets/connection_ssh/test_ssh_askpass.yml (hunk 18 above)
Comment (sivel) [reply test/integration/targets/connection_ssh/test_ssh_askpass.yml:6]:
```
$ ansible-test shell -v --remote macos/15.3
[...]
root@administrators-Mac-3:~/ansible# launchctl stop com.openssh.sshd
Not privileged to stop service.
```

This works on the macos/14.3 remote, and works on my local `15.0.1` that I continuously forget to update :)

**Comment by webknjaz** [comment]
> > > macOS is failing even after rerun
> > 
> > 
> > That failure may be somewhat legitimate. The same setup targets I am using were recently modified. I'll look into it more.
> 
> Looks like it is a change in macos 15, maybe even 15.3, preventing stopping sshd. I'll need to evaluate a change for macos. I tested against the previous macos/14.3 remote and the tests work as expected.

https://discussions.apple.com/thread/255762865 mentions some problems. One is claimed to be fixed on 15.0.1. And there's somebody complaining about 15.2 (file permission bugs?). Plus there are some complaints from Feb.

**Review by nitzmahone — APPROVED** [review]
Works for me as-is. Whether now or later, might be good to change the SHM format to a serialized JSON dict or something to allow for arbitrary other config to be passed over. Later is fine, but we probably should do it before sshpass support actually gets fully killed off to be able to handle the prompt customization as we do now.

**Comment by sivel** [comment]
> Whether now or later, might be good to change the SHM format to a serialized JSON dict or something to allow for arbitrary other config to be passed over

I've transititioned the shm to store JSON for future expansion. I've not made any changes to allow specification or restriction to any particular prompts, I'll log a feature for that once this merges.

Right now it just holds `{"password":"s3cr3t-password-here"}`

File: lib/ansible/cli/_ssh_askpass.py [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+        if HOST_KEY_RE.search(sys.argv[1]):
+            sys.stdout.buffer.write(b'no')
+            sys.stdout.flush()
+            sys.exit(0)
+    except IndexError:
+        pass
+
+    kwargs: dict[str, bool] = {}
+    if sys.version_info[:2] >= (3, 13):
+        # deprecated: description='unneeded due to track argument for SharedMemory' python_version='3.13'
```
Comment (mattclay) [inline lib/ansible/cli/_ssh_askpass.py]: Looking at the conditional here, it seems this deprecation comment was intended to take effect once we drop support for Python 3.12. If that's the case, it should reference that version.

Comment (sivel) [reply lib/ansible/cli/_ssh_askpass.py]: Addressed, good catch.

File: lib/ansible/cli/__init__.py (hunk 19 above)
Comment (webknjaz) [reply lib/ansible/cli/__init__.py]: Are you saying that it *can* be invoked as `python -m ansible` and `argv[0]` would be `'smth.py'`?

```console
$ python -V
Python 3.12.9

$ cat cmd.py
#! /usr/bin/env python
import sys
print(f'{sys.argv[0]=}')
print(f'{sys.argv=}')

$ python -m cmd
sys.argv[0]='/home/usr/full/absolute/path/to/cmd.py'
sys.argv=['/home/usr/full/absolute/path/to/cmd.py']

$ ./cmd.py
sys.argv[0]='./cmd.py'
sys.argv=['./cmd.py']
```

Could you rephrase so it doesn't imply that something doesn't work?

Comment (webknjaz) [reply lib/ansible/cli/__init__.py]: Or did you mean it's the calling side that isn't allowed to have whitespaces in the invocation for some reason?

File: test/integration/targets/connection_ssh/test_ssh_askpass.yml (hunk 18 above)
Comment (webknjaz) [reply test/integration/targets/connection_ssh/test_ssh_askpass.yml:6]: :thinking: I think local would be different since it wouldn't require full-disk access over SSH, though. When I was updating macOS hosts on MacStadium to newer versions, I was also hitting privilege problems post-reboot that required their stuff to physically walk to the machines and use the local keyboard-level access to enable the respective privileges.

File: lib/ansible/cli/__init__.py (hunk 19 above)
Comment (webknjaz) [reply lib/ansible/cli/__init__.py]: I think I understand now. The `SSH_ASKPASS` variable is expected to contain a path and wouldn't contain args. That's probably what you mean here. Still, would be good to spell it out.

File: lib/ansible/plugins/connection/ssh.py [hunk 21]
```diff
@@ -907,6 +944,50 @@ def _examine_output(self, source: str, state: str, b_chunk: bytes, sudoable: boo
...
+        self.shm = shm = SharedMemory(create=True, size=16384, **kwargs)  # type: ignore[arg-type]
+
+        data = json.dumps(
+            {'password': conn_password},
+        ).encode('utf-8')
+        shm.buf[:len(data)] = bytearray(data)
+        shm.close()
+
+        env['_ANSIBLE_SSH_ASKPASS_SHM'] = str(self.shm.name)
+        adhoc = pathlib.Path(sys.argv[0]).with_name('ansible')
```
Comment (webknjaz) [inline lib/ansible/plugins/connection/ssh.py:975]: @sivel why can't this just go to a separate script along the lines of
```suggestion
        adhoc = pathlib.Path(sys.argv[0]).with_name('_ansible_ssh_askpass')
```
?
It would've entirely eliminated the need to go through `lib/ansible/cli/__init__.py` with all that questionable mode detection logic, and would likely only need a new script to be exposed via `pyproject.toml`.

Also, installed scripts could be found through `sys.executable` then, which would be reliable in terms of using the exact same interpreter:
```suggestion
        adhoc = pathlib.Path(sys.executable).with_name('_ansible_ssh_askpass')
```

Comment (mattclay) [reply lib/ansible/plugins/connection/ssh.py:975]: The goal was to avoid adding an entry point just for this feature.

Comment (mkrizek) [reply lib/ansible/plugins/connection/ssh.py:975]: > why can't this just go to a separate script

I was wondering about this too.

> The goal was to avoid adding an entry point just for this feature.

Makes sense but it appears it isn't obvious so I would mention this decision somewhere in a comment, maybe in `lib/ansible/cli/__init__.py` where the detection logic is?

Comment (webknjaz) [reply lib/ansible/plugins/connection/ssh.py:975]: Honestly, I feel like trying to avoid an additional entry point only complicates maintenance. Are we worried that end-users might call it? That'd be detectable through the missing env var.

Comment (sivel) [reply lib/ansible/plugins/connection/ssh.py:975]: If this becomes problematic, we'll re-evaluate adding a new entrypoint. When we discussed this a while back, the complaint was that we have too many entrypoints already, and we'd like fewer.

I'll try to be more verbose in the cli code as to what it's purpose is.
